pyop2/backends/cuda.py

```diff
@@ -0,0 +1,625 @@
+                with self.vec as v:
+                    v.array_w
+                return (self._data.ctypes.data, )
+        else:
+            if cuda_backend.offloading:
+                self.ensure_availability_on_device()
+
+                self._availability_flag = AVAILABLE_ON_DEVICE_ONLY
+                return (self._cuda_data.gpudata, )
+            else:
+                self.ensure_availability_on_host()
+
+                self._availability_flag = AVAILABLE_ON_HOST_ONLY
+                return (self._data.ctypes.data, )
+
+    @mpi.collective
+    @property
+    def data(self):
+        if self.dataset.total_size > 0 and self._data.size == 0 and self.cdim > 0:
+            raise RuntimeError("Illegal access: no data associated with this Dat!")
+        self.halo_valid = False
+
+        if cuda_backend.offloading:
+
+            self.ensure_availability_on_device()
+
+            # {{{ marking data on the host as invalid
+
+            if self.can_be_represented_as_petscvec:
+                # report to petsc that we are updating values on the device
+                with self.vec as petscvec:
+                    petscvec.restoreCUDAHandle(petscvec.getCUDAHandle("w"))
+            else:
+                self._availability_flag = AVAILABLE_ON_DEVICE_ONLY
+
+            # }}}
+            v = self._cuda_data[:self.dataset.size]
+        else:
+            self.ensure_availability_on_host()
+            v = self._data[:self.dataset.size].view()
+            v.setflags(write=True)
+
+            # {{{ marking data on the device as invalid
+
+            if self.can_be_represented_as_petscvec():
+                # report to petsc that we are altering data on the CPU
+                with self.vec as petscvec:
+                    petscvec.array_w
+            else:
+                self._availability_flag = AVAILABLE_ON_HOST_ONLY
+
+            # }}}
+
+        return v
+
+    @property
+    @mpi.collective
+    def data_with_halos(self):
+        self.global_to_local_begin(RW)
+        self.global_to_local_end(RW)
+        return self.data
+
+    @property
+    @mpi.collective
+    def data_ro(self):
+
+        if self.dataset.total_size > 0 and self._data.size == 0 and self.cdim > 0:
+            raise RuntimeError("Illegal access: no data associated with this Dat!")
+
+        self.halo_valid = False
+
+        if cuda_backend.offloading:
+            self.ensure_availability_on_device()
+            v = self._cuda_data[:self.dataset.size].view()
+            # FIXME: PyCUDA doesn't support 'read-only' arrays yet
+        else:
+            self.ensure_availability_on_host()
+            v = self._data[:self.dataset.size].view()
+            v.setflags(write=False)
+
+        return v
+
+    @property
+    @mpi.collective
+    def data_ro_with_halos(self):
+        self.global_to_local_begin(READ)
+        self.global_to_local_end(READ)
+        return self.data_ro
```

Comment:
Tried to replicate what's already implemented (cf https://github.com/OP2/PyOP2/blob/e1f759863e240a7faa609a355646f69b364ab776/pyop2/types/dat.py#L193-L209)

Reply:
I think this is wrong. If you compare the existing implementations of `data_ro` and `data_ro_with_halos` you can see that if we exclude the halos then we only return a slice of `self._data`. This is because the halo data lives at the end of the array.

```py
def data_ro(self):
    v = self._data[:self.dataset.size].view()
    v.setflags(write=False)
    return v

...

def data_ro_with_halos(self):
    v = self._data.view()
    v.setflags(write=False)
    return v
```

Reply:
Aah, I see. Fixed in https://github.com/OP2/PyOP2/compare/3f834d3ae3ac9479792ec78cb984cf95d5121901..5810a4186216fafbbfeb67e074c2f065a5241f02.

Thanks!